frontend/server/src/main/java/org/pytorch/serve/util/GPUManager.java

```diff
@@ -0,0 +1,159 @@
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class GPUManager {
+
+    private static final Logger logger = LoggerFactory.getLogger(GPUManager.class);
+
+    private static GPUManager instance;
+
+    private int nGPUs;
+    private int minFreeMemory;
+    private float maxShareFailures;
+    private AtomicInteger[] nFailures;
+    private AtomicInteger[] freeMemory;
+    private ConcurrentHashMap<String, Integer> workerIds;
+
+    private GPUManager(int nGPUs, int minFreeMemory, float maxShareFailures) {
+        this.nGPUs = nGPUs;
+        this.minFreeMemory = minFreeMemory;
+        this.maxShareFailures = maxShareFailures;
+
+        if (nGPUs > 0) {
+            this.workerIds = new ConcurrentHashMap<String, Integer> ();
+            this.nFailures = new AtomicInteger[this.nGPUs];
+            this.freeMemory = new AtomicInteger[this.nGPUs];
+            for (int i = 0; i < this.nGPUs; i++) {
+                this.nFailures[i] = new AtomicInteger(0);
+                this.freeMemory[i] = new AtomicInteger(-1);
+            }
+        }
+    }
+
+    // code largely copied from WorkerThread::getGpuUsage
+    private int queryNvidiaSmiFreeMemory(int gpuId) {
+        Process process;
+        try {
+            process =
+                    Runtime.getRuntime()
+                            .exec(
+                                    "nvidia-smi -i "
+                                            + gpuId
+                                            + " --query-gpu=memory.free --format=csv,noheader,nounits");
+            process.waitFor();
+            int exitCode = process.exitValue();
+            if (exitCode != 0) {
+                InputStream error = process.getErrorStream();
+                for (int i = 0; i < error.available(); i++) {
+                    logger.error("" + error.read());
+                }
+                return -1;
+            }
+            InputStream stdout = process.getInputStream();
+            BufferedReader reader =
+                    new BufferedReader(new InputStreamReader(stdout, StandardCharsets.UTF_8));
+            String line = reader.readLine();
+            if (line == null) {
+                return -1;
+            } else {
+                return Integer.parseInt(line);
+            }
+        } catch (Exception e) {
+            logger.error("Exception raised : " + e.toString());
+        }
+
+        return -1;
+    }
+
+    public static void init(ConfigManager configManager) {
+        int nGPUs = configManager.getNumberOfGpu();
+        int minFreeMemory = configManager.getMinFreeGpuMemory();
+        float maxShareFailures = configManager.getMaxShareGpuFailures();
+        instance = new GPUManager(nGPUs, minFreeMemory, maxShareFailures);
+    }
+
+    public static GPUManager getInstance() {
+        return instance;
+    }
```

Comment:
When calling `getInstance()` we could also check if `instance` has already been initialized, then we won't have require the user to call `init` manually.

Reply:
Thanks! I will leave this as is for now so the schema remains similar to `ConfigManager`, but we could consider refactoring them both in the future.